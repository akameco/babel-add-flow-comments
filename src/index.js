// @flow
/* ::
type Comment = {
  [key: string]: any,
  value: string,
  ignore: boolean,
}

type Comments = Array<Comment>
*/

const FLOW_DIRECTIVE = '@flow'

export function removeFlowComment(comments /* : Comments */) {
  if (!Array.isArray(comments)) {
    throw new TypeError(`Expected a Array<Comment>, got ${typeof comments}`)
  }

  for (const comment of comments) {
    if (comment.value.indexOf(FLOW_DIRECTIVE) >= 0) {
      comment.value = comment.value.replace(FLOW_DIRECTIVE, '')
      if (!comment.value.replace(/\*/g, '').trim()) {
        comment.ignore = true
      }
    }
  }
}

// line ? /* @flow */ or // @flow
export function addFlowComment(path /* : Object */) {
  path.addComment('leading', ` ${FLOW_DIRECTIVE}`, true)
}
