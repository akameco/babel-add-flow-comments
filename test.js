// @flow
import pluginTester from 'babel-plugin-tester'
import { removeFlowComment, addFlowComment } from './src'

const removePlugin = () => {
  return {
    name: 'remove-comments',
    visitor: {
      Program(path: Object, { file }: Object) {
        // remove @flow
        removeFlowComment(file.ast.comments)
      },
    },
  }
}

const addPlugin = () => {
  return {
    name: 'add-comments',
    visitor: {
      Program(path: Object) {
        addFlowComment(path)
      },
    },
  }
}

pluginTester({
  plugin: removePlugin,
  snapshot: true,
  tests: {
    'remove comments': '// @flow',
  },
})

pluginTester({
  plugin: addPlugin,
  snapshot: true,
  tests: {
    'add comments': '"use strict";',
    func: 'function sum(a, b) { return a + b; }',
    var: 'const a = 2',
  },
})
