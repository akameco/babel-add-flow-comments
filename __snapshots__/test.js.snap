// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. add comments 1`] = `
"
\\"use strict\\";

      ↓ ↓ ↓ ↓ ↓ ↓

// @flow
\\"use strict\\";
"
`;

exports[`1. remove comments 1`] = `
"
// @flow

      ↓ ↓ ↓ ↓ ↓ ↓


"
`;

exports[`2. func 1`] = `
"
function sum(a, b) { return a + b; }

      ↓ ↓ ↓ ↓ ↓ ↓

// @flow
function sum(a, b) {
  return a + b;
}
"
`;

exports[`3. var 1`] = `
"
const a = 2

      ↓ ↓ ↓ ↓ ↓ ↓

// @flow
const a = 2;
"
`;
